src/app/services/study-set-firebase.service.ts

```diff
@@ -26,15 +29,23 @@ export class StudySetFirebaseService extends StudySetService{
   }
 
   override saveStudySet(studySet: StudySetModel): Observable<string> {
-    const docRef = doc(collection(this.firestore, 'study-sets'))
+    const docRef = studySet.id ?
+      doc(this.firestore, 'study-sets', studySet.id) : doc(collection(this.firestore, 'study-sets'));
+    // firebase does not accept custom objects, so must turn into array of structs
+    const sequences = studySet.sequences.map((seq) => ({
+      id: seq.id,
+      name: seq.name,
+      cardList: seq.cardList.map(card => Object.assign({}, card))
+    }))
+    console.log(sequences)
     return defer(() => from(setDoc(docRef, {
       id: docRef.id,
       owner: studySet.owner,
       title: studySet.title,
       description: studySet.description,
-      flashcards: studySet.flashcards,
-      sequences: studySet.sequences
+      flashcards: studySet.flashcards.map(obj => Object.assign({}, obj)),
+      sequences: sequences
     }) as Promise<void> ))
-      .pipe(map(() => studySet.id = docRef.id));
+      .pipe(map(() => studySet.id = docRef.id), tap(thingy => thingy));
```

Comment:
I believe that the `tap(thingy => thingy)` is unneeded, though I do believe that after learning more about the tap function that my use of map was incorrect and should probably be a tap. I am confused about the returning of undefined though. Even though saveStudySet would always generate a new id, it should have at least returned that id. Do you know how you were getting the undefined value? When doing some tests myself using `console.log` to print out values, I would get undefined values sometimes when I expected values that was caused by the asyncronicity of the calls. Depending on how you were testing, the undefined values could have been a result of this.

Reply:
Yes, the tap is unnecessary. I put it there to use the IDE's type sensing. I just didn't delete it when I was done.

I can't recreate the undefined set id bug so I just reverted that portion of the code. I might have put my console log outside the subscribe block when I was testing or I might have tried to call the set ID before it was pushed to the firestore. IDK bro

TBH I didn't put much effort/thought into these changes because I was just trying to get the minimum functioning code that would run on the hosted website. We shouldn't merge this into main without considerable changes and discussion. I just wanted to bring up the issues I ran into  and note some of the bandaids I had to apply to the code to get it to work.